### PR TITLE
[WIP] Draft of cleaner Xonfig loading.

### PR DESCRIPTION
**Pull request by Carreau**
Need to take care of tracking which xonfig have been loaded and don't
reload if the case.

Addresses (in part) #1550 

--- 
Open to discussion.

### CODE REVIEW

File: xonsh/ptk/shell.py [hunk 1]
```diff
@@ -39,6 +39,9 @@ def __init__(self, **kwargs):
         self.key_bindings_manager = KeyBindingManager(**key_bindings_manager_args)
         load_xonsh_bindings(self.key_bindings_manager)
 
+        self._prompt_tokens = None
```
Comment (gforsyth) [inline xonsh/ptk/shell.py]: Hey @Carreau -- re the `_prompt_tokens` stuff, I agree with the changes but is it part of this feature?

Comment (Carreau) [inline xonsh/ptk/shell.py]: No sorry, miss-commited.

**Comment by gforsyth** [comment]
I like this idea -- it's especially nice for completion xontribs, since the `completer add` functionality isn't available until after startup, so some wacky stuff ends up being used to properly load new completions.  

I can't get it to work, though.  I tried to use `xontrib_init` to do a `completer add` but it doesn't get loaded into my completer list.

**Comment by Carreau** [comment]
> I can't get it to work, though. I tried to use xontrib_init to do a completer add but it doesn't get loaded into my completer list.

We likely need to move the xontrib initialisation even later

**Comment by Carreau** [comment]
> @astronouth7303 said: Should this be a dunder special function eg __init__()?

Well, i think that `__init__` would be confusing for various reasons:

1) it is rarely if ever at module scope.
2) It has a specific meaning for Python himself
3) A user coming across this has ~0 change to google the right thing.
4) if you are giving a call and said to your audience to create  `dunder-init` they will be confused. (assuming your users know the "dunder" word.)

if we want a magic method at module level I would thinkmore about `__main__`, but then that makes the module executable.

I'm fine changing but I think the explicit method name is better.

**Comment by scopatz** [comment]
Thanks for putting this in @Carreau! And sorry for taking so long to review this....

In general, I think a named xontrib API is a slippery slope. I think it can become completely unmanageable as needs and use cases change and evolve.  A prime example of this is Vistrails where the API is practically unusable, while being essential to actually writing plugins.  Furthermore, special names should be special, so something like `__xontrib_init__()` would be better than a public variable name to me.

I would prefer some kind of metaprogramming technique.  Maybe a decorator that simply labels the functions to be run (for the time being) would avoid the special naming issue and also allow us to evolve the backend implementation without the xontrib authors needing to worry about a changing API. This could let us swap in an event system later transparently.  For example, I would see something like:

```python
from xonsh.xontrib import xontrib_init

@xontrib_init
def my_initer(*args, **kwargs):
    pass
```

This would even allow a xontrib to have multiple inits, if it needed them for some reason.  Or we could have a more robust system that has different loaders/unloaders

```python
from xonsh.xontribs import loader, unloader

@loader
def my_initer(*args, **kwargs):
    pass

@loader
def my_other_init(*args, **kwargs):
   pass

@unloader
def teardown(*args, **kwargs):
    pass
```

I think this kind of system is a bit more robust overall.  Feel free to reject any names, etc. The main point is that something closer to an event system interface I think is much better than special names.

Thanks again, @Carreau!

**Comment by gforsyth** [comment]
I'm going to close this since we have an events system now.  Please reopen if desired!

**Comment by astronouth7303** [comment]
Not exactly, but this sounds a lot like the ideas for #1550.

**Comment by gforsyth** [comment]
got it, reopening then.

**Comment by astronouth7303** [comment]
No, closed is good. The loading system should go through the event system (as I outlined over there), but there's missing infrastructure pieces before it can start being used.

**Pull request closed without merging**